### PR TITLE
Use full paths when launching processes on Windows

### PR DESCRIPTION
**Pull request by auneri**
The change is related to the [subprocess PATH semantics and portability](https://bugs.python.org/issue8557) and addresses issue #3763.

### CODE REVIEW

**Comment by melund** [comment]
I wonder why I was never able to reproduce the issue for my self. But enough people have reported it... I think the fix looks good. 

Especially if it also smooth out some of the difference  between *nix and windows induced by popen. 

Can you also add a news item so the changelog gets generated correctly on next release.

**Comment by auneri** [comment]
Thank you, added the news item. A few questions though:
1. Would it be better to apply the change for all systems (not just for Windows) for consistency?
2. Is it safe to modify `self.cmd`? I could avoid it if not.
3. Is this proper use of xonsh's `which` command or is there a better way?
4. Not sure why some tests are failing...

**Comment by auneri** [comment]
See my questions above; is there anything else I can provide?

**Comment by melund** [comment]
Sorry. I haven't had time. We definitely shouldn't change the behavior on other than Windows. I think it is OK to use which like you do. 

However, I took a quick look at the tests. Some seem related to this fix. So maybe there are some side effects to this fix. I guess we need to get some one else to take a look at this as well.

**Comment by auneri** [comment]
Anything I can do to help move this forward?
From what I can tell, the failing test is trying to execute `cat tttt | wc` on Windows, which is failing because `cat` is not a valid command.

**Comment by anki-code** [comment]
@auneri try to merge main branch into your branch.

**Comment by auneri** [comment]
Thanks for the suggestion, hope I did it correctly (merged from `xonsh:main`). The Windows errors seem to be unchanged from before.

**Comment by anki-code** [comment]
@auneri Looks like the tests that fail are related to your changes. Click to the "Details" link and fix it.

> which is failing because cat is not a valid command.

These tests were working before your changes, right? This means that the error is unrelated to `cat`.

**Comment by auneri** [comment]
Thanks. I will dig a bit further, but FYI, I am not able to replicate the failures on my local Windows machine. The tests pass regardless of these changes.

**Comment by auneri** [comment]
I was trying to use `xonsh.xoreutils.which` before, but switching to the standard `shutil.which` fixed the tests.

Note also that I specify the path via `builtins.__xonsh__.env.get("PATH")`, omitting that would require `$UPDATE_OS_ENVIRON=True` for paths to resolve properly.

**Comment by auneri** [comment]
Let me know if there is anything I can do on my end. Would love to have this in the next release.

**Comment by anki-code** [comment]
Hi @auneri!

1. We have `resolve_binary_loc` function near your changes which uses `commands_cache.locate_binary`. Could you please check this function and may be it what you need and could be reused?

2. It will be good to have at least one small test for this new logic.

Thanks!

**Comment by auneri** [comment]
1. Alright, turns out the full binary path was already made available via `resolve_binary_loc`; latest change uses that.
2. I will need some guidance on testing. Here is the logic that is corrected via this change (#3763):
```python
python -c "import sys; print(sys.executable)"  # outputs C:\Miniconda3\envs\test\python.exe
$PATH.insert(0, r'C:\Miniconda3\envs\py37')
# before fix
python -c "import sys; print(sys.executable)"  # outputs C:\Miniconda3\envs\test\python.exe
# after fix
python -c "import sys; print(sys.executable)"  # outputs C:\Miniconda3\envs\py37\python.exe
```
Setting `$UPDATE_OS_ENVIRON = True` also corrects this behavior, but only for executables besides `python.exe`. The only way to fix launching correct `python.exe` is with this change.

**Comment by anki-code** [comment]
Ordinary I go to `xonsh/tests` directory and find the appropriate file with tests for module I changed. Then I can run i.e.:
```python
pytest tests/procs/test_specs.py
```
After this I add the new tests or I review the tests around to understand more testing patterns to write new test.

**Comment by auneri** [comment]
A new test in [test_specs.py](https://github.com/xonsh/xonsh/blob/main/tests/procs/test_specs.py) would be appropriate, but the only existing test there is not designed to run on Windows. Looking around a bit more, I was not able to identify a test that is similar to my case, with most tests skipping on Windows.

At this time, my day work is catching up with me and I'm afraid I will not be able to spend more time on this... I would greatly appreciate if someone picks up the mantle.

**Comment by daniel-shimon** [comment]
Hey @anki-code, would you like to pick this up and add a test?

**Comment by anki-code** [comment]
Sorry I'm not the windows user to properly test this )

**Comment by melund** [comment]
I have testet this, and I certainly think we shoud merge it. @daniel-shimon /@anki-code Any objections?

**Comment by daniel-shimon** [comment]
@melund it L9274 good but we need a test in the repo :)
Shouldn't be too complicated, just mock the PATH to a local directory and make sure it calls a correct binary inside

**Comment by melund** [comment]
Do we have some related tests anywhere? I.e. where should we place it?

**Comment by daniel-shimon** [comment]
Probably in the `proc`or `specs` test files

**Comment by melund** [comment]
> Probably in the `proc` or `specs` test files

I ended up with adding them to `test_integrations`. I tried to follow the approach in `test_spec` but I kept getting deadlocks when using `run_subproc` directly. I probably haven't understod how it works directly.

Anyway, I think this should be just as fine. 

Now, I just need to figure out why tests are failing again :-/

**Comment by daniel-shimon** [comment]
You should probably rebase in main first :)

File: tests/test_integrations.py [hunk 1]
```diff
@@ -813,3 +819,36 @@ def test_loading_correctly(monkeypatch, interactive):
...
     assert rtn == 0
+
+
+@skip_if_on_unix
+def test_run_currentfolder():
+    """Ensure we can run an executable in the current folder
+    when file is not on path
+    """
+    batfile = Path(__file__).parent / "bin" / "hello_world.bat"
+    os.chdir(batfile.parent)
```
Comment (daniel-shimon) [inline tests/test_integrations.py:830]: Use the monkeypatch fixture with `monkeypatch.chdir`

**Review by daniel-shimon — CHANGES_REQUESTED** [review]

File: tests/test_integrations.py (hunk 1 above)
Comment (daniel-shimon) [reply tests/test_integrations.py:830]: It's probably the reason other tests are failing

Comment (melund) [reply tests/test_integrations.py:830]: Yes. Let me just try that. Thanks

**Comment by melund** [comment]
Thanks, @daniel-shimon. Tests pass now. You better merge it.

**Review by daniel-shimon — APPROVED** [review]
Great! Thanks @auneri @melund @anki-code for pushing this!

**Pull request merged**